onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +265,128 @@
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(FreeGPUAllocation_2, _In_ void* ptr) {
+  API_IMPL_BEGIN
+#ifdef USE_DML
+  Dml::FreeGPUAllocation(ptr);
+#endif  // USE_DML
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_DML2, _In_ OrtSessionOptions* options, OrtDmlDeviceOptions* device_opts) {
+API_IMPL_BEGIN
+  //options->provider_factories.push_back(onnxruntime::DMLProviderFactoryCreator::Create(device_id));
+  
+    OrtDmlPerformancePreference p = device_opts->p;
+    OrtDmlDeviceFilter f = device_opts->f;
+
+    // Create DXCore Adapter Factory
+    ComPtr<IDXCoreAdapterFactory> adapterFactory;
+    ORT_THROW_IF_FAILED(::DXCoreCreateAdapterFactory(adapterFactory.GetAddressOf()));
+
+    // Get a list of all the adapters that support compute
+    ComPtr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
```

Comment:
>d3D12CoreComputeAdapters [](http://example.com/codeflow?start=31&length=24)

You need to call the sort api with the appropriate preference before iterating on it.

Reply:
Is this still valid? Doesn't it make sense to filter the devices first, then sort them?

Reply:
You cannot call the AdapterList::Sort API with an array of adapters you construct your self. You need to call it on the AdapterList object. In addition, you cannot construct an adapter list from an arbitrary list of adapters.

You have to do it first.